Major refactor, new features, testing

garrett92895 opened the pull request:
Hello!

Let me start off by acknowledging that this is a giant PR. Long story short, shortly after creating #40 I assumed the repo was dead and started making plans to create my own, similar add-on. I started refactoring and adding features without worrying about ease of merging/review. Let me know how you'd like to approach this.

## Additions
1. Added a nice default styling to the cards. Example:
<img width="250" alt="Screenshot 2025-06-24 at 16 51 43" src="https://github.com/user-attachments/assets/cc517ecb-31eb-464e-ad58-133a1a9f4918" />

2. Improved syncing capabilities. The sync now has a checkbox to allow the user to also downgrade the status in lingq. Additionally, there is much more robust error handling and retry logic, as well as a progress bar. 
<img width="410" alt="Screenshot 2025-06-24 at 16 59 20" src="https://github.com/user-attachments/assets/19fdcfa6-5840-4e32-9d95-7cb9bf775341" />
<br></br>
<img width="316" alt="Screenshot 2025-06-24 at 16 57 39" src="https://github.com/user-attachments/assets/34706db4-3b16-48de-acaa-866085f0f8c6" />

3. Extensive testing, including a functioning mock Lingq API and a mock `mw` object, along with test data taken from a real lingq account and a real anki database

4. Card template comes with an empty FrontAudio and SentenceAudio to allow the user to populate the cards with a text-to-speech tool such as AwesomeTTS
 
5. Added a check on language code. Checks user input against all current LingQ language codes. Might need to be updated as lingq adds more languages

6. Added logic to limit the number of cards we try to sync to lingq


## Changed
1. Refactor how we handle the mapping between lingq status/extended_status and anki. This was done using an enum, of sorts, inside the `Lingq` model class along with some helper methods. Also, the card no longer stores `Status` and `ExtendedStatus` but a simple string such as "recognized" (using lingq parlance) inside of a field `LingqStatus` for simplicity 
2. The fields on the anki note type are slightly different now
3. I changed the default `status_to_interval` list. I wanted it so that when you hit "Easy" in Anki, it's a guaranteed increase on the status (assuming a ease factor of 2.5, anki's default)
4. Made `AnkiCard` and `Lingq` dataclasses
5. General refactoring

## What's left
1. Add a live progress bar to import. Right now there is only a progress bar on syncing since that seemed to take a lot longer than importing, and also to be more prone to 429 responses from the API.
2. Most values that we store in the `AnkiCard` are two-way, meaning they can be converted back into a `Lingq` that is nearly identical to its original value (straight from the API). However, the exception is with the `translations` field, since we add some html styling on the `AnkiCard`. Ideally when we convert from a `AnkiCard` to `Lingq` object, we would have some logic to split the translations 
i.e.
```
AnkiCard: {
  "translations": ["translation1<br>translation2"]
}
->
LingQ: {
  "translations": ["translation1", "translation2"]
}
```
however, this hasn't caused any issues as of yet.
3. As noted in the Future Goals section of the readme, I added some things like allowing a user-defined status_to_interval, logging, etc.

garrett92895 committed:
Fix 404 errors from LingQ API and add popularity attribute

garrett92895 committed:
Update requirements

garrett92895 committed:
Fix interval bug

garrett92895 committed:
working logic

garrett92895 committed:
update comment

garrett92895 committed:
Add button to sync only unkown linqgs

garrett92895 committed:
Merge branch 'import-knowns-checkbox' into interval_fix

garrett92895 committed:
Pythonify code

garrett92895 committed:
Input checking language code

garrett92895 committed:
Better format and styling for cards

garrett92895 committed:
Allow sync to downgrade lingqs

garrett92895 committed:
Update back template

garrett92895 committed:
rename package back to forked original

garrett92895 committed:
Retry requests when timeout is hit

garrett92895 committed:
black

garrett92895 committed:
Remove redundant exception

kovyrin committed:
Give import_knowns a default value (#42)

garrett92895 committed:
Add black github action

garrett92895 committed:
Fix worflow

garrett92895 committed:
Create test data files for proper mocking

garrett92895 committed:
remove uneeded test files

garrett92895 committed:
Test coverage for converter and import fixing

garrett92895 committed:
Mocks for anki handler and lingq api

garrett92895 committed:
black

garrett92895 committed:
Finish testing

garrett92895 committed:
Remove unecessary TODO

garrett92895 committed:
Add progress bar to syncing

garrett92895 committed:
Update README

garrett92895 committed:
Rename everything to old casing

garrett92895 committed:
rename directory

garrett92895 committed:
Remove old folder

garrett92895 committed:
Remove black styling

garrett92895 committed:
rename file

garrett92895 committed:
Sync 2.0

garrett92895 committed:
Merge

garrett92895 committed:
Delete duplicate file

garrett92895 committed:
Fix init

garrett92895 committed:
Casing of mw

garrett92895 committed:
Undo renaming of anki addon

garrett92895 committed:
Codacy suggestions

garrett92895 committed:
Codacy

garrett92895 committed:
codacy again

garrett92895 commented:
@thags I've been using this version of the addon myself for the past few months (on a mac) and everything seems quite stable. Additional user testing would definitely be helpful. One particular issue I could see coming up is since I changed the fields on the anki cards, this extension will no longer be compatible with decks made using the old version of the addon.

thags commented:
@garrett92895. First off thanks for all the details and screenshots, and general work done. 
It is indeed huge, preferably it would be broken into smaller chunks with a pr for each independent feature/refactor. Looking at the commit history I think that would be a lot of work to break off like that. Or it would create commits that depend on other commits which would be worse than this mega commit. 

I'll plan to just work through the review bit by bit. It may take a while.

I'll keep in mind the potential break for current users.

thags commented on LingqAnkiSync/AnkiHandler.py:
```diff
@@ -42,18 +57,39 @@ def CreateNoteType(name: str, fields: List):
     template = mw.col.models.newTemplate("lingqAnkiSync")
-    template['qfmt'] = "{{Front}}<br>{{Sentence}}"
-    template['afmt'] = "{{FrontSide}}<hr id=answer>{{Back}}"
+    resourceFolder = os.path.dirname(__file__) + "/resources"
```
This will likely break on windows. 
```suggestion
    resourceFolder = os.path.join(os.path.dirname(__file__), "resources")
```

garrett92895 replied:
Oh good catch 👍

thags commented on LingqAnkiSync/AnkiHandler.py:
```diff
@@ -42,18 +57,39 @@ def CreateNoteType(name: str, fields: List):
-    template['afmt'] = "{{FrontSide}}<hr id=answer>{{Back}}"
+    resourceFolder = os.path.dirname(__file__) + "/resources"
+
+    with open(resourceFolder + "/style.css", "r") as f:
```
```suggestion
    with open(os.path.join(resourceFolder + "style.css"), "r") as f:
```

thags commented on LingqAnkiSync/AnkiHandler.py:
```diff
@@ -42,18 +57,39 @@ def CreateNoteType(name: str, fields: List):
+    with open(resourceFolder + "/style.css", "r") as f:
+        model["css"] = f.read()
+
+    with open(resourceFolder + "/front.html", "r") as f:
```
```suggestion
    with open(os.path.join(resourceFolder + "front.html"), "r") as f:
```

thags commented on LingqAnkiSync/AnkiHandler.py:
```diff
@@ -42,18 +57,39 @@ def CreateNoteType(name: str, fields: List):
+    with open(resourceFolder + "/front.html", "r") as f:
+        template["qfmt"] = f.read()
+
+    with open(resourceFolder + "/back.html", "r") as f:
```
```suggestion
    with open(os.path.join(resourceFolder + "back.html"), "r") as f:
```

thags commented on LingqAnkiSync/AnkiHandler.py:
```diff
@@ -42,18 +57,39 @@ def CreateNoteType(name: str, fields: List):
+    card.note()["LingqStatus"] = status
+    mw.col.update_note(card.note())
+
+    # Anki seems to miss a few of them if the updates aren't spaced out. This isn't a perfect solution
+    time.sleep(0.1)
+
```
Reference: https://addon-docs.ankiweb.net/the-anki-module.html

We could build the updates needed into a csv and then import the csv so update all the cards at once. Thus not needing this sleep. I am not certain that would work. 

If not, then we could interact with the database directly and call executemany().

thags replied:
I am okay with leaving the time.sleep for now. This can be a future improvement.

thags commented on LingqAnkiSync/resources/back.html:
```diff
@@ -0,0 +1,17 @@
+        <p class="example-sentence-audio example-sentence">"{{Sentence}}"{{SentenceAudio}}</p>
+    </div>
+    <div class="back">
+        <a href="https://context.reverso.net/translation/spanish-english/{{Front}}">Reverse Context</a>
```
This context.revoerso should not be a default, as it only applies in spanish-english scenarios, or more work should be done to generalize it. To any language to any language scenario.

garrett92895 replied:
You're right, I've now generalized it in ff2d44e to work with any of the supported languages on context.reverso

thags commented on Tests/data/anki_data.json:
Is there a reason there is so much data for test? Should we not condense it to a smaller subset? For both this and lingq_data.json.

garrett92895 replied:
There's not a strong reason for it. The test data is basically an export of my real anki deck and lingq vocab, which I deliberately didn't sync for a week or so. I wanted the test data to represent a real-world, slightly out-of-sync pair of datasets without having to edit the test data by hand to make it smaller.

However, it really isn't that much work to make a smaller subset and keep that slightly out-of-sync property that it has, and it would definitely make the tests run faster. I'll make that change soon

garrett92895 replied:
Actually now that I think about it some more, I wanted a larger dataset so that we could make sure that paging is tested. `page_size=200` is hard-coded in `GetLingqs`

thags replied:
The mock api could just copy data 200 times for the response. And even though the pagesize is hardcoded to 200, that is a maximum, not a minimum. The last page received will almost always be fewer than 200.

It feels like we are tightly coupling our tests to lingqs current api, a specific language code, and a specific state in time. Then we are really testing the implementation of our mocks instead of our logic. I think it could be decoupled from these and require a lot less of an initial dataset.

And like you said, it would speed up the time to run the tests. Right now it is looping over every item in the dataset one by one to retrieve a specific item, which is really slow. We could improve that speed by other methods, but the most straight forward would to just have fewer items.

These are my current thoughts, let me know if you think I am off base. I will look over the tests a few more times to get a clearer view.

garrett92895 replied:
I see what you're saying. My approach was to model as close to a real-world scenario as possible, but in doing so I sort of conflated unit testing with this case-specific integration testing.

I'm happy to shrink the test set and rewrite the tests so they test things at a closer level. I think much of the slow speed comes from the mock server serving up HTTP responses, and from the time.sleep calls (which can also simply be mocked).

My question is how would you recommend going from here. Should I remove the mock server / `mock_anki.py` or simply shrink the test set and restructure what I have?

I'm happy to do whatever, it was an interesting exercise regardless.

thags replied:
I would remove the mock anki and mock server. In the end testing that way will create a higher maintenance burden, and does not help us if for example lingq changes their api. It may actual confuse us since the tests still work, but not in the real world.

Then the testing data can be trimmed down a lot to only use what is needed for specific examples. Like converting between the api and anki. If certain things can not be tested without calling the api or anki then we should look at splitting out the logic to separate functions if we want to test, instead of creating these mocks.

garrett92895 replied:
Okay I've implemented this feedback into `test_AnkiHandler.py` as of cff5215, which no longer uses mockMw. If this file looks more in line with what you were thinking then I can update the rest of the tests similarly, and then delete the mockMw and mock server.

thags replied:
Yes, that looks in line with what I was thinking. Thank you!

thags commented on Tests/conftest.py:
```diff
@@ -0,0 +1,44 @@
+
+@pytest.fixture
+def lingqData():
+    with open("Tests/data/lingq_data.json") as f:
```
All of the open statements in this file should use os.path.join

thags commented on LingqAnkiSync/Converter.py:
```diff
@@ -1,67 +1,117 @@
+        raise ValueError(f'No such status "{status}". Should be one of {Lingq.LEVELS}')
+
+    extendedStatus = 0
+    if status == "known":
```
```suggestion
    if status == Lingq.LEVEL_KNOWN:
```

thags commented on LingqAnkiSync/Converter.py:
```diff
@@ -1,67 +1,117 @@
+) -> List[Lingq]:
     lingqs = []
     for card in ankiCards:
+        status, extendedStatus = LingqStatusToInternalStatus(
+            _AnkiIntervalToLingqStatus(card.interval, statusToInterval)
+        )
```
I think the outer function "LingqStatusToInternalStatus" is named inappropriately. It is really converting our internal lingq status (lingq.Level) to external lingq status (3,0). 

And the flow feels off with needing to call _AnkiIntervalToLingqStatus, which sounds more appropriate as the main function. 

I would propose a new function that is called like this:
```
status, extendedStatus = LingqLevelToLingqApi(card.interval, statusToInterval)
```

garrett92895 replied:
Yea I struggled with this. They say one of the hardest things about programming is naming things and I definitely felt that here 😅

The idea is to completely abstract away (status+extendedStatus) and anki interval as much as possible, so that most of the app just deals with a simple form ('new', 'known', etc.) which here I basically to as either `LingqStatus`/ `knownStatus`. I thought about naming them `canonicalStatus` and `lingqInternalStatus`, or something similar. 

What that would mean is basically

```
---- Current code ----
status, extendedStatus = CanonicalStatusToLingqInternalStatus(
    _AnkiIntervalToCanonicalStatus(card.interval, statusToInterval)
)

---- Proposed changes ----
status, extendedStatus = AnkiIntervalToLingqInternalStatus(card.interval, statusToInterval)
```

Which is decidedly wordy, but perhaps worth it.

We could also use the naming convention of lingqLevel ('new', 'known', etc.) and lingqStatus (status+extendedStatus), but I feel like that can get confusing as well.

As for the proposed flow changes, I'm happy to write it either way, to me the important thing is keeping consistent naming and maintaining the abstraction to keep `(status, extendedStatus)` and `interval` at the "edges" of the code

Let me know your thoughts

thags replied:
I agree that there probably is not a perfect naming solution that is not confusing. Your idea for lingqLevel and lingqStatus seems fine to me if it is used consistently. That should also shorten the names a bit, instead of `AnkiStatusToLingqInternalStatus` it could be `AnkiToLingqStatus`

garrett92895 replied:
Ok I've standardized the naming in 25835d1

1. 'new', 'recognized', etc. exclusively referred to as "level"
2. `(status, extendedStatus)` consistently referred to as `lingqStatus` 
  a. "status" alone refers to the first value in `(status, extendedStatus)`)
3. `AnkiInterval` now just called `Interval`

Much more readable now IMO as there are no overloaded uses of the word "status" 👍

thags commented on LingqAnkiSync/Converter.py:
```diff
@@ -1,67 +1,117 @@
+        )
+
+    if extendedStatus == 3:
+        knownStatus = Lingq.LEVELS[4]
```
```suggestion
        knownStatus = Lingq.LEVEL_KNOWN
```

garrett92895 committed:
Platform-agnostic file paths

garrett92895 committed:
Fix references to lingq levels

thags commented:
Thinking about the compatibility for anyone on a previous version:
Do you think a conversion class could be made? Where we detect if someone is on an old version, probably by if their card contains an old field. If detected then it will convert those items to the new way?

garrett92895 committed:
Remove test notetype naming

garrett92895 committed:
Different templates for different languages

garrett92895 committed:
Standardize wording for 'status'

garrett92895 committed:
Make codacy happy

garrett92895 committed:
Small boolean logic change

garrett92895 commented:
> Thinking about the compatibility for anyone on a previous version:
Do you think a conversion class could be made? Where we detect if someone is on an old version, probably by if their card contains an old field. If detected then it will convert those items to the new way?

@thags I'm trying to think of the simplest way to handle it with minimal tech debt. I feel like adding logic in the UIHandler (or elsewhere) to automatically check which version the user is on (and automatically migrate it) is going to be messy and difficult to test. I'd almost rather let the app fail if the user's deck is on the old schema, but then at the same time we add a button to the UI with something like "Click Here to Upgrade Notes to V2". Since the old schema has `LingqStatus` and `LingqExtendedStatus`, it should be straightforward to map to the new `LingqLevel` field. 

I was previously worried about the app silently failing with this approach, since this PR originally repurposed the `LingqStatus` field and simply dropped `LingqExtendedStatus`. However, now that we've refactored it so that the card has `LingqLevel` as a field instead, I'm more confident that users on the old version will not be able to run any of the new functions until they hit the "Upgrade Notes to V2" button

Thinking out loud, I think the challenges I think are going to be
1. how easy is it with anki's `mw` to change a notetype schema
2. how easy will it be to update the template/styling 
3. handling renaming of the notetype/template, since we are now using a languageCode suffix in the notetype name (e.g. convert `lingqAnkiSync` to `lingqAnkiSync_es`)

thags commented:
> > Thinking about the compatibility for anyone on a previous version:
> > Do you think a conversion class could be made? Where we detect if someone is on an old version, probably by if their card contains an old field. If detected then it will convert those items to the new way?
> 
> @thags I'm trying to think of the simplest way to handle it with minimal tech debt. I feel like adding logic in the UIHandler (or elsewhere) to automatically check which version the user is on (and automatically migrate it) is going to be messy and difficult to test. I'd almost rather let the app fail if the user's deck is on the old schema, but then at the same time we add a button to the UI with something like "Click Here to Upgrade Notes to V2". Since the old schema has `LingqStatus` and `LingqExtendedStatus`, it should be straightforward to map to the new `LingqLevel` field.
> 
> I was previously worried about the app silently failing with this approach, since this PR originally repurposed the `LingqStatus` field and simply dropped `LingqExtendedStatus`. However, now that we've refactored it so that the card has `LingqLevel` as a field instead, I'm more confident that users on the old version will not be able to run any of the new functions until they hit the "Upgrade Notes to V2" button
> 
> Thinking out loud, I think the challenges I think are going to be
> 
> 1. how easy is it with anki's `mw` to change a notetype schema
> 2. how easy will it be to update the template/styling
> 3. handling renaming of the notetype/template, since we are now using a languageCode suffix in the notetype name (e.g. convert `lingqAnkiSync` to `lingqAnkiSync_es`)

I agree it would be a lot of work to handle the conversion. I could see a few ways to do it that would not be so bad, but probably not worth the time because Lingq acts as kind of a backup, since the two should be synced the data should not be that outdated.

As long as their is something that tells the user what steps they need to take, like create a new deck to sync to then I think that is enough.

garrett92895 committed:
Rewrite AnkiHandler tests

garrett92895 committed:
Renaming

garrett92895 committed:
Refactor LingqApi tests

garrett92895 committed:
Refactor UI Action Handler tests

garrett92895 committed:
Formatting and codacy suggestions

garrett92895 commented:
Okay @thags I feel we are almost there!

### Tests
I've now refactored my tests and removed the mockMw and mock server as well as the big testing datasets.

### NoteType breaking changes
I've done some simulating of how the addon will behave when using this new version on a deck created by the old version. The main takeaway is that we are going to need to implement _some_ amount of version-aware code, because as-is it will actually successfully import new cards, but it does so by creating a new noteType, and then you end up with a deck with two different noteTypes. On syncing it does in fact fail, but with a very unhelpful message that says nothing except `'LingqLevel'`. The good news is that even with the new noteType, importing does not create duplicates in the deck.

Anyway, here are my ideas. 

**1. Simplest technically but worst UX**
We add some code in the beginning of import/sync to see if the user is on the old version of the addon (by looking for LingQStatus and LingQExtendedStatus on the cards) and then raise an exception telling the user to start fresh with a new deck and re-import from LingQ. Paired with a clear message on AnkiWeb, this seems fair, however if the user hasn't synced their cards to LingQ _before_ installing the updated addon then it's going to be a pain for them.

**2. Best UX but most complicated**
Add a migration functionality to addon. Either with a button ("Click Here to upgrade deck") or automatically. Honestly, this seems like it could be quite error-prone and messy.

**3. Middle ground**
We allow multiple noteTypes in the deck and only really pay attention to the ones with the correct noteType. This only really affects syncing. Once the user installs the new version of the addon, their existing cards will now have the "wrong" noteType and those cards would be ignored for syncing, but any new cards that they import from lingq will work with syncing. I think this is actually simple to implement, and the UI could have a message like "Synced 50 cards and ignored 10 (wrong noteType)" because as it is there isn't really anything stopping people from mixing noteTypes in the deck anyway. Of course, we should also leave a note of this in ankiweb or in the addon so people understand why it's behaving that way.

I also looked at ways the user could manually map the old noteType to the new one using Anki's built in tools, but unfortunately it looks like it needs to be done through code (or another addon).

Let me know your thoughts, if you have a preference, or any ideas of your own!

thags commented:
Awesome work @garrett92895 ! 

Any amount of code that we add to migrate or detect old versions is maintenance overhead being added. I would prefer to keep that to a minimum, but there is a balance to be had with making sure users at least understand any common errors.

I like your third option, it is a bit more generalized.

garrett92895 committed:
Backwards compatibility

garrett92895 commented:
Ok should be all set! I've tested this on my end with a mix of versions and it works fine. 

<img width="282" height="247" alt="Screenshot 2025-07-23 at 19 10 29" src="https://github.com/user-attachments/assets/71cdbac6-3ece-412d-8a5d-34ae9659528a" />

Users should still be careful not to put multiple types of notes into the deck, for example by putting multiple LingQ languages in one deck.

I also noticed a funny quirk while testing, which is that if you're in the midst of being rate-limited by LingQ while syncing, you actually can't use the website at the same time either 😅

Let me know if there's any other thoughts you have but otherwise thanks for all the feedback and working through this with me!

garrett92895 committed:
typo

thags commented:
Thank @garrett92895 I will merge this in now. It will stay take some time before I upload it to ankiweb since I want to put some info about the migration.
